source/sigv4.c

```diff
@@ -947,17 +1236,515 @@ static SigV4Status_t verifySigV4Parameters( const SigV4Parameters_t * pParams )
+    assert( pHmacContext->pCryptoInterface->hashFinal != NULL );
+
+    pCryptoInterface = pHmacContext->pCryptoInterface;
+
+    /* Write the inner hash. */
+    returnStatus = pCryptoInterface->hashFinal( pCryptoInterface->pHashContext,
+                                                innerHashDigest,
+                                                pCryptoInterface->hashDigestLen );
+
+    if( returnStatus == 0 )
+    {
+        /* Create the outer-padded key by retrieving the original key from
+         * the inner-padded key then XOR with opad. XOR is associative,
+         * so one way to do this is by performing XOR on each byte of the
+         * inner-padded key with (0x36 ^ 0x5c) = (ipad ^ opad) = 0x6a.  */
+        for( i = 0U; i < pCryptoInterface->hashBlockLen; i++ )
+        {
+            pHmacContext->key[ i ] ^= ( char ) ( 0x6a );
+        }
+
+        returnStatus = pCryptoInterface->hashInit( pCryptoInterface->pHashContext );
+    }
+
+    if( returnStatus == 0 )
+    {
+        /* Update hash using the outer-padded key. */
+        returnStatus = pCryptoInterface->hashUpdate( pCryptoInterface->pHashContext,
+                                                     pHmacContext->key,
+                                                     pCryptoInterface->hashBlockLen );
+    }
+
+    if( returnStatus == 0 )
+    {
+        /* Update hash using the inner digest. */
+        returnStatus = pCryptoInterface->hashUpdate( pCryptoInterface->pHashContext,
+                                                     innerHashDigest,
+                                                     pCryptoInterface->hashDigestLen );
+    }
+
+    if( returnStatus == 0 )
+    {
+        /* Write the final HMAC value. */
+        returnStatus = pCryptoInterface->hashFinal( pCryptoInterface->pHashContext,
+                                                    pMac,
+                                                    macLen );
+    }
+
+    /* Reset the HMAC context. */
+    pHmacContext->keyLen = 0U;
+
+    return returnStatus;
+}
+
+static SigV4Status_t writeLineToCanonicalRequest( const char * pLine,
+                                                  size_t lineLen,
+                                                  CanonicalContext_t * pCanonicalContext )
+{
+    SigV4Status_t returnStatus = SigV4Success;
+
+    if( pCanonicalContext->bufRemaining < lineLen + LINEFEED_CHAR_LEN )
+    {
+        LogError( ( "Insufficient space in processing buffer for canonical request. "
+                    "Increase `SIGV4_PROCESSING_BUFFER_LENGTH` to fix, bytesExceeded=%zu.",
+                    lineLen - pCanonicalContext->bufRemaining ) );
+        returnStatus = SigV4InsufficientMemory;
+    }
+    else
+    {
+        ( void ) memcpy( pCanonicalContext->pBufCur,
+                         pLine,
+                         lineLen );
+        pCanonicalContext->pBufCur += lineLen;
+
+        *pCanonicalContext->pBufCur = LINEFEED_CHAR;
```

Comment:
I'm also going to point out again that this won't work if the length is not 1, so you might as well just use 1 instead of `LINEFEED_CHAR_LEN`.
```suggestion
        *( pCanonicalContext->pBufCur ) = LINEFEED_CHAR;
```